lib/eof.go

```diff
@@ -13,13 +14,14 @@ var yamuxSessionShutdown = yamux.ErrSessionShutdown.Error()
 // IsErrEOF returns true if we get an EOF error from the socket itself, or
 // an EOF equivalent error from yamux.
 func IsErrEOF(err error) bool {
-	if err == io.EOF {
+	if errors.Is(err, io.EOF) {
 		return true
 	}
 
 	errStr := err.Error()
 	if strings.Contains(errStr, yamuxStreamClosed) ||
-		strings.Contains(errStr, yamuxSessionShutdown) {
+		strings.Contains(errStr, yamuxSessionShutdown) ||
+		strings.HasSuffix(errStr, io.EOF.Error()) {
```

Comment:
In theory this seems reasonable, but I think because `EOF.Error()` is such a short string (`EOF`), there is a pretty significant chance this incorrectly matches a a different error.

A suffix of `: EOF` might be a bit safer, but I'm not sure that is necessary either. I think we should be encouraging all code to use wrapping and not rely on string comparison. Ideally we would remove the need for these other `strings.Contains()` as well.

I think we should not add this `strings.HasSuffix`, and fix any code that does not properly wrap errors to do so.

Reply:
I agree on principle and long-term that's what i'd want too.

To give some context on the string comparison, the code that motivated it is here: 
https://github.com/hashicorp/net-rpc-msgpackrpc/blob/master/client.go#L33

In this case, the error comes in the Response.Error field as a string, and the string is copied into an error object. In the specific case I am seeing the string is: `rpc error: EOF`. 
This error created by this line of code does not unwrap correctly as an EOF and as far as I can tell, cannot be identified as an EOF other than by string comparison. Maybe there's a quick fix I don't know about and if that's the case I'll happily submit another PR for it, but if there isn't i'm afraid the bigger fix involves changing how the RPC client works, and If I understand the situation correctly, this is used in many places in many hashicorp projects. Which makes this change really scary for a first time external contributor. 

Thoughts?

Reply:
I really like this idea, I'll send an update.

The scenario i'm testing is on long-poll requests where a user of the API is connected to a nomad client, the client is forwarding the RPC to a server, and while the request is blocked, the leader is killed. In that specific case, sometimes, I see this wrapped error as string. I think it may have to do with how the requests are forwarded between the client, its server, and potentially the cluster leader. I'm still learning how that codepath works, I've only been staring at it for two days :)

What can I do that would help narrow this down? I can repro on a fairly regular basis, I just need help orienting myself in the code and adding logs where we need them to confirm the problematic code path(s).

Reply:
I have added the ServerError type assertion: I hope it is restrictive enough and solves at least one of the failure mode I'm trying to get past (explained below, step 3). I'm not out of the woods yet but if this gets checked in, I hope it'll help improve robustness. 

**Long explanation**
I am starting to understand that the problem I'm trying to solve goes deeper and lies in the forwarding chain of the RPC requests in a nomad cluster. Here's the scenario:

1. api user makes a blocking request (wait on an index) -> nomad client -> nomad server (leader)
2. leader dies/is killed while the request is waiting. Returns an EOF error
3. *with this fix*: client retries and changes server: client -> nomad server(follower) (without the fix, the client would not understand it received an EOF error)
4. nomad server (follower) -> nomad server(leader) : OOPS the leader is down: "connection refused" error (not EOF and the leader election hasn't happened yet)
5. follower does not retry and passes the error (connection refused) down to the client in the `response.Error` field
6. client: not an EOF, returns a wrapped version of the connection refused to the caller <- that's my new problem :)

From the caller perspective this is difficult, if not impossible, to handle correctly: the caller shouldn't be expected to understand (and recover from) RPC forwarding chain errors. There are a finite number of scenarios like this (depending whether the client is connected to a follower or a leader, how fast leader election goes, and which server goes down, the one doing the forwarding or the leader). I think this should be entirely handled within the RPC chain between the nomad client and servers. retries and HA are hard problems though :) 

This being said, I don't think it's realistic to expect that all cases can be covered with `IsErrEOF()`, which is already a little bit extended past EOF (the yamuxError, and now some ServerErrors provided they end with `: EOF`). The follow up investigation is probably directly in the nomad RPC code.

In conclusion, I still think this is a net improvement and we've pushed this PR as far as it should go, happy to make more changes as required to get it merged, this is a move in the right direction :)